### PR TITLE
[asl][typing] Typing Reference Pre-Beta step 3

### PR DESCRIPTION
**Pull request by Roman-Manevich**
Work to get the ASL Typing Reference ready for another round of reviews on May 13th.
- [x] Introduction and background.
- [x] Domain of Values for Types 
- [x] Basic Type Attributes
- [x] Relations Over Types
- [x] Typing of bitfields
- [x] Typing of Types
- [x] Typing of Expressions
- [x] Typing of Left-Hand-Side Expressions
- [x] Typing of Slices
- [x] Typing of Patterns
- [x] Typing of Local Declarations
- [ ] Typing of Statements
- [x] Typing of Blocks
- [x] Typing of Catchers
- [ ] Typing of Subprogram Calls
- [x] Typing of Subprograms
- [ ] Typing of Global Declarations
- [ ] Typing of Specifications
- [ ] Static Evaluation
- [ ] Symbolic Subsumption Testing
- [ ] Symbolic Equivalence Testing
- [ ] Utility Relations
- [ ] Address existing feedback.

### CODE REVIEW

File: asllib/doc/ASLTypingReference.tex [hunk 1]
```diff
@@ -4575,12 +4632,12 @@ \section{TypingRule.EGetBadRecordField \label{sec:TypingRule.EGetBadRecordField}
   The result of annotating the expression $\ve$ in $\tenv$ is
 $(\vt, \newe)$ and all of the following apply:
   \begin{itemize}
-  \item $\ve$ denotes the access of field \texttt{field\_name} on expression $\veone$;
-  \item \texttt{t\_e1, e2} is the result of annotating $\veone$ in $\tenv$;
-  \item \texttt{t\_e1} has the structure of an exception or record type with fields $\fields$;
-  \item \texttt{t\_e2} has the structure of an exception or record type with fields $\fields$;
-  \item \texttt{t\_e2} is an Exception or a Record type with fields $\fields$;
-  \item \texttt{field\_name} is not declared in $\fields$;
+  \item $\ve$ denotes the access of field $\fieldname$ on expression $\veone$;
```
Comment (jalglave) [inline asllib/doc/ASLTypingReference.tex:4635]: On expression?

File: asllib/doc/ASLTypingReference.tex [hunk 2]
```diff
@@ -4602,10 +4659,10 @@ \section{TypingRule.EGetBadBitField \label{sec:TypingRule.EGetBadBitField}}
   The result of annotating the expression $\ve$ in $\tenv$ is
 $(\vt, \newe)$ and all of the following apply:
   \begin{itemize}
-  \item $\ve$ denotes the access of field \texttt{field\_name} on expression $\veone$;
-  \item \texttt{t\_e1} has the structure a bitvector type with bitfields $\bitfields$;
-  \item \texttt{t\_e2} has the structure a bitvector type with bitfields $\bitfields$;
-  \item \texttt{field\_name} is not declared in $\bitfields$;
+  \item $\ve$ denotes the access of field $\fieldname$ on expression $\veone$;
```
Comment (jalglave) [inline asllib/doc/ASLTypingReference.tex:4662]: On expression?

File: asllib/doc/ASLTypingReference.tex [hunk 3]
```diff
@@ -4627,9 +4684,9 @@ \section{TypingRule.EGetBadField \label{sec:TypingRule.EGetBadField}}
   The result of annotating the expression $\ve$ in $\tenv$ is
 $(\vt, \newe)$ and all of the following apply:
    \begin{itemize}
-   \item $\ve$ denotes the access of field \texttt{field\_name} on expression $\veone$;
-   \item \texttt{t\_e1, e2} is the result of annotating $\veone$ in $\tenv$;
-   \item \texttt{t\_e1} does not have the structure of a record or an exception or a bitvector type;
+   \item $\ve$ denotes the access of field $\fieldname$ on expression $\veone$;
```
Comment (jalglave) [inline asllib/doc/ASLTypingReference.tex:4687]: On expression?

File: asllib/doc/ASLTypingReference.tex [hunk 4]
```diff
@@ -4651,11 +4708,11 @@ \section{TypingRule.EGetBitField \label{sec:TypingRule.EGetBitField}}
   The result of annotating the expression $\ve$ in $\tenv$ is
 $(\vt, \newe)$ and all of the following apply:
   \begin{itemize}
-  \item $\ve$ denotes the access of field \texttt{field\_name} on expression $\veone$;
-  \item \texttt{t\_e1, e2} is the result of annotating $\veone$ in $\tenv$;
-  \item the \underlyingtype\ of \texttt{t\_e1} is a bitvector type with bitfields $\bitfields$;
-  \item \texttt{field\_name} is declared in $\bitfields$;
-  \item $\slices$ gives the slices corresponding to the bitfield \texttt{field\_name}
+  \item $\ve$ denotes the access of field $\fieldname$ on expression $\veone$;
```
Comment (jalglave) [inline asllib/doc/ASLTypingReference.tex:4711]: On expression?

File: asllib/doc/ASLTypingReference.tex [hunk 5]
```diff
@@ -4688,15 +4745,15 @@ \section{TypingRule.EGetBitFieldNested \label{sec:TypingRule.EGetBitFieldNested}
   The result of annotating the expression $\ve$ in $\tenv$ is
 $(\vt, \newe)$ and all of the following apply:
   \begin{itemize}
-  \item $\ve$ denotes the access of field \texttt{field\_name} on expression $\veone$;
-  \item \texttt{t\_e1, e2} is the result of annotating $\veone$ in $\tenv$;
-  \item \texttt{t\_e1} has the structure of a bitvector type with bitfields $\bitfields$;
-  \item \texttt{field\_name} is declared in $\bitfields$;
-  \item $\slices$ gives the slices corresponding to the bitfield \texttt{field\_name} in \\
+  \item $\ve$ denotes the access of field $\fieldname$ on expression $\veone$;
```
Comment (jalglave) [inline asllib/doc/ASLTypingReference.tex:4748]: On expression?

File: asllib/doc/ASLTypingReference.tex [hunk 6]
```diff
@@ -4753,6 +4809,30 @@ \section{TypingRule.EGetBitFieldTyped \label{sec:TypingRule.EGetBitFieldTyped}}
 
 \isempty{\subsection{Comments}}
 
+\section{TypingRule.EGetTupleItem \label{sec:TypingRule.EGetTupleItem}}
+
+\subsection{Prose}
+All of the following apply:
```
Comment (jalglave) [inline asllib/doc/ASLTypingReference.tex:4815]: Do we need the itemize here? It looks like there’s only one clause?

File: asllib/doc/ASLTypingReference.tex [hunk 7]
```diff
@@ -4753,6 +4809,30 @@ \section{TypingRule.EGetBitFieldTyped \label{sec:TypingRule.EGetBitFieldTyped}}
...
+\begin{itemize}
+  \item $\ve$ denotes \texttt{e1, field\_name};
+\end{itemize}
+
+\subsection{Example}
```
Comment (jalglave) [inline asllib/doc/ASLTypingReference.tex:4820]: Here, and in fact everywhere: we need small representative examples like in the semantics reference, which should also be used for regression.

File: asllib/doc/ASLTypingReference.tex [hunk 8]
```diff
@@ -5756,73 +5871,94 @@ \section{TypingRule.LEConcat \label{sec:TypingRule.LEConcat}}
...
+A type error is returned, if one is detected.
 
-  \subsection{Prose}
-   Annotating a slice~\texttt{slice} in an environment $\tenv$
-(\texttt{annotate\_slice env slice}) results in the pair \texttt{(offset, length)} and all of the following apply:
-   \begin{itemize}
-   \item \texttt{slice} gives an index \texttt{i};
-   \item \texttt{(offset, length)} is the result of applying TypingRule.SliceLength to \texttt{i, i+:1}.
-   \end{itemize}
+The relevant rule is given by:
```
Comment (jalglave) [inline asllib/doc/ASLTypingReference.tex:5904]: Why the forward pointer sorry?

File: asllib/doc/ASLTypingReference.tex [hunk 9]
```diff
@@ -6234,64 +6421,55 @@ \chapter{Typing of Local Declarations}
...
-   \item \texttt{new\_ldi} is the declaration of variable $\vx$ with type $\tty$.
-   \end{itemize}
-
-  \subsection{Example}
-    \VerbatimInput{../tests/ASLTypingReference.t/TypingRule.LDVar.asl}
+\subsection{Prose}
+All of the following apply:
+\begin{itemize}
+  \item $\ldi$ denotes a variable $\vx$, that is, $\LDIVar(\vx)$;
+  \item determing whether $\vx$ is not declared in $\tenv$ yields $\True$ \ProseOrTypeError;
```
Comment (jalglave) [inline asllib/doc/ASLTypingReference.tex:6454]: Typo: determing

File: asllib/doc/ASLTypingReference.tex [hunk 10]
```diff
@@ -6300,37 +6478,34 @@ \section{TypingRule.LDVar \label{sec:TypingRule.LDVar}}
...
-    \subsection{Example}
-      \VerbatimInput{../tests/ASLTypingReference.t/TypingRule.LDTyped.asl}
+\subsection{Prose}
+All of the following apply:
+\begin{itemize}
+  \item $\ldi$ denotes a local declaration item $\ldip$ with local declaration keyword $\ldk$
+  and a type $\vt$, that is $\LDITyped(\ldip, \vt)$;
+  \item annotating the type $\vt$ in $\tenv$ yields $\vtp$ \ProseOrTypeError;
+  \item determining whether $\vtp$ can be initialized with $\tty$ in $\tenv$ yields $\True$ \ProseOrTypeError;
+  \item anotating the local declaration item $\ldip$ with the local declaration keyword $\ldk$, given
```
Comment (jalglave) [inline asllib/doc/ASLTypingReference.tex:6488]: Typo: anotating

File: asllib/doc/ASLTypingReference.tex [hunk 11]
```diff
@@ -6339,40 +6514,40 @@ \section{TypingRule.LDTyped\label{sec:TypingRule.LDTyped}}
...
+\subsection{Prose}
+All of the following apply:
+\begin{itemize}
+  \item $\ldi$ denotes a tuple of local declaration items $\ldi_{1..k}$, that is, $\LDITuple(\ldi_{1..k})$;
+  \item determining the \structure\ of $\tty$ in $\tenv$ yields $\vtp$ \ProseOrTypeError;
+  \item determining whether $\vtp$ is a tuple type with $k$ elements yields $\True$ \ProseOrTypeError;
+  \item annotating the local declaration items in $\ldis$ from right to left with their corresponding
+  (that is, with the same index) types $t_{1..k}$ in $\tenv$,
+  propagating static environments from one annotation to the next,
+  yields the local delcaration items $\ldip_{1..k}$ \ProseOrTypeError;
```
Comment (jalglave) [inline asllib/doc/ASLTypingReference.tex:6526]: Typo: delcaration

**Review by jalglave — COMMENTED** [review]
Thanks @Roman-Manevich - a few minor comments.

File: asllib/doc/ASLTypingReference.tex [hunk 12]
```diff
@@ -5179,6 +5259,41 @@ \subsection{Formally}
...
+annotates a statically-evaluable integer expression $\ve$ in the static environment $\tenv$
+and returns the annotated expression $\vepp$.
+A type error is returned, if one is detected.
+
+\subsection{Prose}
+All of the following apply:
+\begin{itemize}
+  \item annotating the expression $\ve$ in $\tenv$ yields $ (\vt, \vep)$ \ProseOrTypeError;
+  \item determining whether $\vt$ is a statically \constrainedinteger\ in $\tenv$ yields $\True$ \ProseOrTypeError;
+  \item determining whether $\vep$ is statically-evaluable in $\tenv$ yields $\True$ \ProseOrTypeError;
```
[the remark this answers is not in the document]
Comment (Roman-Manevich) [reply asllib/doc/ASLTypingReference.tex:5279]: fixed

File: asllib/doc/ASLTypingReference.tex (hunk 6 above)
Comment (Roman-Manevich) [reply asllib/doc/ASLTypingReference.tex:4815]: the list is not done. I need to go back and fill it.

File: asllib/doc/ASLTypingReference.tex (hunk 7 above)
Comment (Roman-Manevich) [reply asllib/doc/ASLTypingReference.tex:4820]: absolutelu. I think that after the beta I'm going to add a lot of non-tricial examples that demonstrate the tricky parts.

File: asllib/doc/ASLTypingReference.tex (hunk 8 above)
Comment (Roman-Manevich) [reply asllib/doc/ASLTypingReference.tex:5904]: I like listing all the rules in the chapter but the previous rules go with the "One of the following applies:" and this rule does not.

File: asllib/doc/ASLTypingReference.tex (hunk 7 above)
Comment (jalglave) [reply asllib/doc/ASLTypingReference.tex:4820]: Fantastic, thank you.

File: asllib/doc/ASLTypingReference.tex (hunk 8 above)
Comment (jalglave) [reply asllib/doc/ASLTypingReference.tex:5904]: Sorry I am not sure I understand. Let’s discuss Monday?

**Pull request merged**